Blocks/URL/url.h

```diff
@@ -361,29 +356,82 @@ struct URLParametersExtractor {
   std::string fragment;
-  std::string url_without_parameters;
 };
 
 struct URL : URLParametersExtractor, URLWithoutParametersParser {
   URL() = default;
 
-  // Extra parameters for previous host and port are provided in the constructor to handle redirects.
-  URL(const std::string& url,
-      const std::string& previous_scheme = kDefaultScheme,
-      const std::string& previous_host = "",
-      const uint16_t previous_port = 0)
-      : URLParametersExtractor(url),
-        URLWithoutParametersParser(
-            URLParametersExtractor::url_without_parameters, previous_scheme, previous_host, previous_port) {}
+  URL(const std::string& url) {
+    size_t parameters_start_index = 0;
+    URLParametersExtractor::ExtractURLParametersFromURL(url, parameters_start_index);
+    URLWithoutParametersParser::ParseURLWithoutParameters(url.substr(0, parameters_start_index));
+  }
 
-  URL(const std::string& url, const URLWithoutParametersParser& previous)
-      : URLParametersExtractor(url),
-        URLWithoutParametersParser(URLParametersExtractor::url_without_parameters, previous) {}
+  std::string ComposeURLWithoutParameters() const {
+    return URLWithoutParametersParser::ComposeURL();
+  }
 
   std::string ComposeURL() const {
     return URLWithoutParametersParser::ComposeURL() + URLParametersExtractor::ComposeParameters();
   }
 
+  static URL MakeURLWithDefaults(const URL& source_url_parsed) {
+    URL url = URL(source_url_parsed);
+
+    if (url.scheme.empty()) {
+      if (url.port == 0) {
+        url.scheme = kDefaultScheme;
+      } else {
+        url.scheme = DefaultSchemeForPort(url.port);
+        if (url.scheme.empty()) {
+          url.scheme = kDefaultScheme;
+        }
+      }
+    }
+
+    if (url.host.empty()) {
+      url.host = kDefaultHost;
+    }
+
+    if (url.port == 0) {
+      url.port = DefaultPortForScheme(url.scheme);
+    }
+
+    return url;
+  }
+
+  static URL MakeRedirectedURL(const URL& source_url_parsed, const URL& target_url_parsed) {
+    URL url = URL(source_url_parsed);
+
+    if (!target_url_parsed.scheme.empty()) {
+      url.scheme = target_url_parsed.scheme;
+    }
+
+    if (!target_url_parsed.host.empty()) {
+      url.host = target_url_parsed.host;
+      url.port = target_url_parsed.port;
+    } else if (target_url_parsed.port > 0) {
+      url.port = target_url_parsed.port;
+    }
+
+    url.path = target_url_parsed.path;
+
+    url.parameters_vector = target_url_parsed.parameters_vector;
+    url.query = target_url_parsed.query;
+    url.fragment = target_url_parsed.fragment;
+
+    return MakeURLWithDefaults(url);
+  }
+
+  static URL MakeRedirectedURL(const URL& source_url_parsed, const std::string& url) {
+    return URL::MakeRedirectedURL(source_url_parsed, URL(url));
+  }
+
+  // Parses query strings without the question mark and `application/x-www-form-urlencoded` HTTP bodies.
+  static const std::map<std::string, std::string> ParseQueryString(const std::string& body) {
```

Comment:
I'm sure we could use it in a few other places, where the `URL("/?blahblahblah")` hack is employed now. Thanks @sompylasar.